finality-aleph/src/metrics/chain_status.rs

```diff
@@ -0,0 +1,239 @@
+use futures::{future, StreamExt};
+use log::warn;
+use sc_client_api::{FinalityNotifications, ImportNotifications};
+use sp_api::{BlockT, HeaderT};
+use sp_blockchain::{lowest_common_ancestor, HeaderMetadata};
+use sp_runtime::{
+    traits::{SaturatedConversion, Zero},
+    Saturating,
+};
+use substrate_prometheus_endpoint::{
+    register, Gauge, Histogram, HistogramOpts, PrometheusError, Registry, U64,
+};
+use tokio::select;
+
+use crate::{metrics::LOG_TARGET, BlockNumber};
+
+enum ChainStatusMetrics {
+    Prometheus {
+        top_finalized_block: Gauge<U64>,
+        best_block: Gauge<U64>,
+        reorgs: Histogram,
+    },
+    Noop,
+}
+
+impl ChainStatusMetrics {
+    fn new(registry: Option<Registry>) -> Result<Self, PrometheusError> {
+        let registry = match registry {
+            Some(registry) => registry,
+            None => return Ok(ChainStatusMetrics::Noop),
+        };
+
+        Ok(ChainStatusMetrics::Prometheus {
+            top_finalized_block: register(
+                Gauge::new("aleph_top_finalized_block", "no help")?,
+                &registry,
+            )?,
+            best_block: register(Gauge::new("aleph_best_block", "no help")?, &registry)?,
+            reorgs: register(
+                Histogram::with_opts(
+                    HistogramOpts::new("aleph_reorgs", "Number of reorgs by length")
+                        .buckets(vec![1., 2., 3., 5., 10.]),
+                )?,
+                &registry,
+            )?,
+        })
+    }
+
+    fn noop() -> Self {
+        ChainStatusMetrics::Noop
+    }
+
+    fn update_best_block(&self, number: BlockNumber) {
+        if let ChainStatusMetrics::Prometheus { best_block, .. } = self {
+            best_block.set(number as u64)
+        }
+    }
+
+    fn update_top_finalized_block(&self, number: BlockNumber) {
+        if let ChainStatusMetrics::Prometheus {
+            top_finalized_block,
+            ..
+        } = self
+        {
+            top_finalized_block.set(number as u64);
+        }
+    }
+
+    fn report_reorg(&self, length: BlockNumber) {
+        if let ChainStatusMetrics::Prometheus { reorgs, .. } = self {
+            reorgs.observe(length as f64);
+        }
+    }
+}
+
+pub async fn start_chain_state_metrics_job_in_current_thread<
```

Comment:
This name is _ever so slightly_ overspecified. I would go with just `run`, but `run_chain_state_metrics` is also alright.

Reply:
True, changed